doc/dirvish-runall.8
.\" Automatically generated by Pod::Man 2.1801 (Pod::Simple 3.07)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "DIRVISH-RUNALL 1"
.TH DIRVISH-RUNALL 1 "2009-02-04" "perl v5.10.0" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
dirvish\-runall \- run a set of dirvish backup jobs.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBdirvish-runall\fR\fI[\s-1OPTIONS\s0]\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Run all the jobs specified in the \fBRunall\fR field of the dirvish master
configuration file.
.SH "OPTIONS"
.IX Header "OPTIONS"
Each option may be unambiguously abbreviated.
.PP
\&\fB\-\-config\fR \fIconfigfile\fR
.PP
Read \fIconfigfile\fR for the \fBRunall:\fR field. By default
\&\fBdirvish-runall\fR will use the \fBRunall:\fR field from the master
configuration file.
.PP
This will only be used by \fBdirvish-runall\fR itself and not be
propagated to the \fBdirvish\fR processes.
.PP
See \fB\f(BIdirvish.conf\fB\|(5)\fR for more details.
.PP
\&\fB\-\-no\-run\fR
.PP
\&\fB\-\-dry\-run\fR
.PP
Don't actually do anything. Just display what would have happened.
.PP
\&\fB\-\-quiet\fR
.PP
Run quietly, only report errors.
.PP
Normally \fBdirvish-runall\fR will report the dirvish commands invoked
with timestamps.
.PP
\&\fB\-\-version\fR
.PP
Print the version information and exit.
.PP
\&\fB\-\-pidfile\fR
.PP
Specify an alternate location for the pidfile.
.SH "EXIT CODES"
.IX Header "EXIT CODES"
To facilitate further automation and integration of \fBdirvish-runall\fR
with other tools \fBdirvish-runall\fR provides rationalised exit codes.
The exit codes are range based. While the code for a specific error may
change from one version to another it will remain within the specified
range. So don't test for specific exit codes but instead test for a
range of values. To the degree possible higher value ranges indicate
more severe errors.
.IP "\fB0\fR" 10
.IX Item "0"
success
.IP "\fB1\-199\fR" 10
.IX Item "1-199"
Number of \fBdirvish\fR jobs that failed. If more than 199 jobs failed 199
will be the exit code.
.IP "\fB200\-219\fR" 10
.IX Item "200-219"
An error was encountered in loading a configuration file.
.IP "\fB220\-254\fR" 10
.IX Item "220-254"
An error was detected in the configuration.
.IP "\fB255\fR" 10
.IX Item "255"
Incorrect usage.
.SH "FILES"
.IX Header "FILES"
\&\fB/etc/dirvish/master.conf\fR
.PP
alternate master configuration file.
.PP
\&\fB/etc/dirvish.conf\fR
.PP
master configuration file.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
.Vb 1
\& dirvish.conf(5)
.Ve
.SH "AUTHOR"
.IX Header "AUTHOR"
Dirvish was created by J.W. Schultz of Pegasystems Technologies.
.PP
Dirvish is now maintained by Keith Lofstrom at www.dirvish.org , with
the able assistance of many others.
.SH "BUGS"
.IX Header "BUGS"


debian/changelog
dirvish (1.3.2-r657-1) unstable; urgency=low

  * New upstream release

 -- tex <tex@iowa.gauner.org>  Sun, 08 Feb 2009 00:41:43 +0100

dirvish (1.3.2-r652-1) unstable; urgency=low

  * New upstream release

 -- tex <tex@iowa.gauner.org>  Wed, 04 Feb 2009 19:34:46 +0100

dirvish (1.3.2-r634-1) unstable; urgency=low

  * New upstream release

 -- tex <tex@iowa.gauner.org>  Wed, 28 Jan 2009 17:46:52 +0100

dirvish (1.3.2-r632-1) unstable; urgency=low

  * Initial release

 -- Dominik Schulz <dominik.schulz@gauner.org>  Tue, 27 Jan 2009 16:18:14 +0100


dirvish-runall.pl
#!/usr/bin/perl
# dirvish-runall
# 1.3.X series
# Copyright 2005 by the dirvish project
# http://www.dirvish.org
#
# Last Revision   : $Rev: 652 $
# Revision date   : $Date: 2009-02-04 19:34:29 +0100 (Mi, 04 Feb 2009) $
# Last Changed by : $Author: tex $
# Stored as       : $HeadURL: https://secure.id-schulz.info/svn/tex/priv/dirvish_1_3_1/dirvish-runall.pl $

#########################################################################
#                                                         				#
#	Licensed under the Open Software License version 2.0				#
#                                                         				#
#	This program is free software; you can redistribute it				#
#	and/or modify it under the terms of the Open Software				#
#	License, version 2.0 by Lauwrence E. Rosen.							#
#                                                         				#
#	This program is distributed in the hope that it will be				#
#	useful, but WITHOUT ANY WARRANTY; without even the implied			#
#	warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR				#
#	PURPOSE.  See the Open Software License for details.				#
#                                                         				#
#########################################################################

#----------------------------------------------------------------------------
# Revision information
my %CodeID = (
    Rev    => '$Rev: 652 $'     ,
    Date   => '$Date: 2009-02-04 19:34:29 +0100 (Mi, 04 Feb 2009) $'    ,
    Author => '$Author: tex $'  ,
    URL    => '$HeadURL: https://secure.id-schulz.info/svn/tex/priv/dirvish_1_3_1/dirvish-runall.pl $' ,
);

$VERSION =   $CodeID{URL};
$VERSION =~  s#^.*dirvish_##;  # strip off the front
$VERSION =~  s#\/.*##;         # strip off the rear after the last /
$VERSION =~  s#[_-]#.#g;       # _ or - to "."

#----------------------------------------------------------------------------
# Modules and includes
#----------------------------------------------------------------------------
use strict;
use warnings;

use Time::ParseDate;
use POSIX qw(strftime);
use Getopt::Long;
use Dirvish;

#----------------------------------------------------------------------------
# SIG Handler
#----------------------------------------------------------------------------
$SIG{TERM} = \&sigterm; # handles "kill <PID>"
$SIG{INT} = \&sigterm; # handles Ctrl+C or "kill -2 <PID>"

#----------------------------------------------------------------------------
# Initialisation
#----------------------------------------------------------------------------
our $VERSION;

my $Options = {
        version         => sub {
                        print STDERR "dirvish version $VERSION\n";
                        exit(0);
                },
        help            => \&usage,
};

$Options = reset_options( \&usage, @ARGV);	# initialize the %$Options hash
load_master_config(undef, $Options);		# load master config into $Options

GetOptions($Options, qw(
        config=s
        pidfile=s
        quiet no-run|dry-run
        version help
    )) or &usage();

$$Options{Dirvish} ||= 'dirvish';

$$Options{'no-run'} and $$Options{quiet} = 0;

my $errors = 0;

if(!check_pidfile($$Options{pidfile})) {
	print "$0 already running. Aborting at ".strftime('%H:%M:%S', localtime)." - PID: ".$$."\n";
	exit 255; # Abort
}

my $thmutex = undef;						# Thread mutex
my @threads = undef;						# List of running threads
if($$Options{'Threads'}) {
	use Thread qw(async);
	use Thread::Semaphore;
	$thmutex = Thread::Semaphore->new($$Options{'Threads'});	# Limit the number of running threads
	@threads = ();
}

for my $sched (@{$$Options{Runall}})
{
	my ($vault, $itime) = split(/\s+/, $sched);
	print "runall - vault: $vault\n";
	my $cmd = "$$Options{Dirvish} --vault $vault";
	$itime and $cmd .= qq[ --image-time "$itime"];
	$$Options{quiet}
		or printf "%s %s\n", strftime('%H:%M:%S', localtime), $cmd;
	if($$Options{'Threads'}) {
		my $t = async {
			$thmutex->down();			# Semaphore will block until we have a free slot
			$$Options{'no-run'} and $cmd .= " --no-run";	# Actually execute dirvish, even when running
			# in no-run mode, but make it also run in no-run mode
			print "Thread ".Thread->self->tid()." running cmd: $cmd\n" unless $$Options{quiet};
		    my $status = system($cmd);
		    $thmutex->up();				# Release semaphore lock so another thread can run
	    	return $status;
	    };
	    push(@threads,$t);
    } else {
    	# Without threads, everything is done sequential
    	my $status = system($cmd) unless $$Options{'no-run'};
	    $status < 0 || $status / 256 and ++$errors;
    }
}

if($$Options{'Threads'}) {
	# Wait for the threads to finish
	for my $th (@threads) {
		my $status = $th->join();
		$status < 0 || $status / 256 and ++$errors;
	}
}

remove_pidfile($$Options{pidfile});

$$Options{quiet}
    or printf "%s %s\n", strftime('%H:%M:%S', localtime), 'done';


exit ($errors < 200 ? $errors : 199);
#----------------------------------------------------------------------------
# Subs
#----------------------------------------------------------------------------
sub usage
{
    my $message = shift(@_);

    length($message) and print STDERR $message, "\n\n";

    $! and exit(255); # because getopt seems to send us here for death

    print STDERR <<EOUSAGE;
USAGE
	dirvish-runall OPTIONS
OPTIONS
	--config configfile
	--no-run
	--quiet
	--version
EOUSAGE

	exit 255;
}
# Remove the pidfile on exit
END {
	# remove the pidfile on exit
	remove_pidfile($$Options{pidfile});
}
# Handle SIGTERM (SIG-15)
sub sigterm
{
	print STDERR "Received SIGTERM. Aborting running backup ...";
	# kill childs - kill(TERM, -$$):
	use POSIX;
	my $cnt = kill(SIGTERM, -$$);
	no POSIX;
	print STDERR "Signaled $cnt processes in current processgroup";
	# quit
	exit;
}
#----------------------------------------------------------------------------
# EOF
#----------------------------------------------------------------------------
